### ISSUE TITLE
Why model_bundle is not predicting custom data?

### ISSUE THREAD

**Issue by Vijay-P1999** [azavea/raster-vision#1396]
## ❓ Questions and Help
---

Hey, I tried training semantic segmentation in RV using [this example](https://docs.rastervision.io/en/0.13/examples.html#semantic-segmentation-spacenet-vegas)

> export RAW_URI="data/spacenet-dataset/" 
> export ROOT_URI="data/local-output/"
> rastervision run local rastervision_pytorch_backend/rastervision/pytorch_backend/examples/semantic_segmentation/spacenet_vegas -a raw_uri $RAW_URI -a root_uri $ROOT_URI

I downloaded the dataset using `aws s3 sync`.The data download failed in the middle, but i started training with the data available.
My dataset directory looks like this -
`/data/spacenet-dataset/spacenet/SN2_buildings/train/AOI_2_Vegas`
 The model trained for 4 epochs. 

And then when i used the `model_bundle.zip` to make predictions on my own data. It didn't predict anything at all.

![image](https://user-images.githubusercontent.com/96561146/165919561-47ed599e-af83-493b-bac6-3b637dbc50c6.png)

**❓Why is this happening? Should i change the number of epochs or another parameters?**

Can the model which was trained using Vegas dataset predict my data? If not, please provide documentation for custom training of semantic segmentation model and labelling, exporting tif data. Thanks in advance.

**Environment:**
Ubuntu 20.04 LTS
> 2022-04-29 08:59:46:rastervision.pytorch_learner.learner: INFO - Devices:
2022-04-29 08:59:46:rastervision.pytorch_learner.learner: INFO - index, name, driver_version, memory.total [MiB], memory.used [MiB], memory.free [MiB]
0, NVIDIA GeForce RTX 2070 SUPER, 510.60.02, 8192 MiB, 1781 MiB, 6192 MiB

> 2022-04-29 08:59:46:rastervision.pytorch_learner.learner: INFO - PyTorch version: 1.9.1+cu102
> 2022-04-29 08:59:46:rastervision.pytorch_learner.learner: INFO - CUDA available: True
> 2022-04-29 08:59:46:rastervision.pytorch_learner.learner: INFO - CUDA version: 10.2
> 2022-04-29 08:59:46:rastervision.pytorch_learner.learner: INFO - CUDNN version: 7605
> 2022-04-29 08:59:46:rastervision.pytorch_learner.learner: INFO - Number of CUDA devices: 1
> 2022-04-29 08:59:46:rastervision.pytorch_learner.learner: INFO - Active CUDA Device: GPU 0

**Comment by AdeelH** [comment]
Can you share `$ROOT_URI/train/test_preds.png`? Also, the metrics achieved at the end of training.

**Comment by Vijay-P1999** [comment]
@AdeelH Here it is, test_preds.png
![test_preds](https://user-images.githubusercontent.com/96561146/165931615-6c6c6ee8-1f72-4c80-8f87-75b92283798e.png)

And I assume, by metrics you mean this logs
![image](https://user-images.githubusercontent.com/96561146/165932569-e8ad4f76-65bf-4339-ae95-6a9244ddf851.png)

**Comment by AdeelH** [comment]
I think I know what is going wrong. Can you open you input image in QGIS and share what its histogram looks like?

**Comment by Vijay-P1999** [comment]
This is how the histogram is @AdeelH ,
![histogram](https://user-images.githubusercontent.com/96561146/165935209-64ce9314-926f-4abd-ab7b-c6100d340e06.png)

If you want some more details, i can mail you my input image. If its ok with you, Pls share the mail id.

**Comment by AdeelH** [comment]
The problem seems to be that the model has been trained on images with very different histograms: 
![image](https://user-images.githubusercontent.com/13014700/165936002-6ed2ee32-1062-48d4-b71d-ff7129c1744e.png)

The solution to this is to use the `StatsTransformer`, which normalizes input images to 0-255, so that the model learns to work with that range of values. This used to be there in that example but we removed it in 0.13 because the model seemed perfectly capable of learning without it. But I guess that has the disadvantage of making it less useful on new imagery.

The solution is to put it back in. To do that you'll need to change the following lines
https://github.com/azavea/raster-vision/blob/0.13/rastervision_pytorch_backend/rastervision/pytorch_backend/examples/semantic_segmentation/spacenet_vegas.py#L100-L101

to 
```python
    raster_source = RasterioSourceConfig(
        uris=[image_uri], 
        channel_order=channel_order, 
        transformers=[StatsTransformerConfig()])
```

And then re-run training. You will notice that this time RV runs an `analyze` stage before `train`. The `analyze` stage determines the statistics of pixel values in the image and these stats are then used to normalize the values.

**Comment by Vijay-P1999** [comment]
Thanks @AdeelH , I'll try the above mentioned steps and update in this issue.

**Comment by Vijay-P1999** [comment]
Hey @AdeelH,
I re-trained the model after adding `StatsTransformerConfig()`. The result is the same black image for my data.

This is what i got - test_preds.png
![test_preds](https://user-images.githubusercontent.com/96561146/166091170-36ca0938-22a4-4482-a43b-e6bd06fc3af5.png)

This model was trained for 5 epochs. Should i train it for more epochs? What do you suggest?

**Comment by AdeelH** [comment]
Can you share the logs too?

**Comment by Vijay-P1999** [comment]
I figured out the problem. I re-trained the model for 10 epochs. Now the model predicts my data.
But the result was not as expected. 
This is the result of the newly trained model
![test_preds](https://user-images.githubusercontent.com/96561146/166677353-abe4f1ab-5225-4565-8aa2-ca6f18baefd8.png)

As you can see from the above image, the predicted areas are not accurate. I guess that can be solved by training the model for few more epochs. 
@AdeelH If you have any suggestions for improving the model for custom data you can mention it, or else we can close this issue.

**Comment by AdeelH** [comment]
> Now the model predicts my data.

That's great! Glad it worked out.

> But the result was not as expected.

Right. What you are trying to do is take a model trained on one dataset and apply it to data outside that dataset ("out of distribution"). This is not, in general, expected to always work perfectly. You can think of it like this: the model has learned to detect the kinds of buildings that are present in Las Vegas from imagery of a specific resolution etc. from a specific satellite. If you feed it more images of the same area from the same satellite, you would expect it to do very well, but as you move away from that configuration (by using a different city or satellite) you should expect the results to be much worse.

Training for longer can certainly help, but you also risk over-fitting your training set. In general, the best way to improve out-of-distribution generalization is to improve your training data. You can do this by making it more diverse e.g. by adding scenes from other places or by using stronger data augmentation.

**Comment by Vijay-P1999** [comment]
Thanks Adeel,
> In general, the best way to improve out-of-distribution generalization is to improve your training data.

Will keep this is mind for further progress. Cheers
